Assertion `a && "computeValue() must have assignment"' failed

# Klee Bug Report

Hi, I encountered a crash bug when using KLEE to test a C program. Could you take a look at this problem? Thank you.

### Operating System： 
Ubuntu 18.04

### Klee Version:
KLEE 2.1-pre (https://klee.github.io)           
Build mode: RelWithDebInfo (Asserts: ON)               
Build revision: 64a2b1a93648461f11daa44911d791c28bc00885                  

LLVM (http://llvm.org/):                
LLVM version 6.0.0                   

Optimized build.                 
Default target: x86_64-pc-linux-gnu               
Host CPU: skylake                   

### Test File         
The following C code has been reduced by C-Reduce.                    
***Test.c***                  

```C
b;
a(c) {
  if (!c)
    abort();
}
main() {
  int e;
  short d;
  klee_make_symbolic(&d, sizeof(d), "__sym___VERIFIER_nondet_short");
  b = d;
  a(b > 0);
  int f[b];
  int g[b];
  for (;;) {
    short d;
    klee_make_symbolic(&d, sizeof(d), "__sym___VERIFIER_nondet_short");
    a(d >= 0 && d < b);
    f[d];
    g[b - 1 - d];
  }
}
```

### Bug description
First compile Test.c to LLVM bitcode.                
`clang -I ../../include -emit-llvm -c -g Test.c`                    
Then run klee on the bitcode file.                  
`klee Test.bc`                   
There will be an Assertion failed.                  


###Error Log
```shell
KLEE: output directory is "/home/quan/CWorkSpace/BUG_TEST/klee-out-0"
KLEE: Using STP solver backend
WARNING: this target does not support the llvm.stacksave intrinsic.
KLEE: ERROR: Test.c:4: abort failure
KLEE: NOTE: now ignoring this error at this location
KLEE: ERROR: Test.c:12: concretized symbolic size
KLEE: NOTE: now ignoring this error at this location
KLEE: ERROR: Test.c:18: memory error: out of bound pointer
KLEE: NOTE: now ignoring this error at this location
klee: /home/quan/klee/klee/lib/Solver/CexCachingSolver.cpp:332: virtual bool CexCachingSolver::computeValue(const klee::Query&, klee::ref<klee::Expr>&): Assertion `a && "computeValue() must have assignment"' failed.
/usr/lib/llvm-6.0/lib/libLLVM-6.0.so.1(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamE+0x2a)[0x7f10031130ea]
/usr/lib/llvm-6.0/lib/libLLVM-6.0.so.1(_ZN4llvm3sys17RunSignalHandlersEv+0x56)[0x7f1003111366]
/usr/lib/llvm-6.0/lib/libLLVM-6.0.so.1(+0x81c49b)[0x7f100311149b]
/lib/x86_64-linux-gnu/libc.so.6(+0x3ef20)[0x7f1001e1bf20]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xc7)[0x7f1001e1be97]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x141)[0x7f1001e1d801]
/lib/x86_64-linux-gnu/libc.so.6(+0x3039a)[0x7f1001e0d39a]
/lib/x86_64-linux-gnu/libc.so.6(+0x30412)[0x7f1001e0d412]
klee(_ZN16CexCachingSolver12computeValueERKN4klee5QueryERNS0_3refINS0_4ExprEEE+0x4ec)[0x5644e247effc]
klee(_ZN17IndependentSolver12computeValueERKN4klee5QueryERNS0_3refINS0_4ExprEEE+0x149)[0x5644e2470e59]
klee(_ZN4klee6Solver8getValueERKNS_5QueryERNS_3refINS_12ConstantExprEEE+0x128)[0x5644e2475438]
klee(_ZN4klee12TimingSolver8getValueERKNS_14ExecutionStateENS_3refINS_4ExprEEERNS4_INS_12ConstantExprEEE+0x264)[0x5644e243b394]
klee(_ZNK4klee12AddressSpace10resolveOneERNS_14ExecutionStateEPNS_12TimingSolverENS_3refINS_4ExprEEERSt4pairIPKNS_12MemoryObjectEPKNS_11ObjectStateEERb+0x1a6)[0x5644e243dc66]
klee(_ZN4klee8Executor22executeMemoryOperationERNS_14ExecutionStateEbNS_3refINS_4ExprEEES5_PNS_12KInstructionE+0x123)[0x5644e24053d3]
klee(_ZN4klee8Executor18executeInstructionERNS_14ExecutionStateEPNS_12KInstructionE+0x1e3e)[0x5644e240949e]
klee(_ZN4klee8Executor3runERNS_14ExecutionStateE+0x71d)[0x5644e240fd7d]
klee(_ZN4klee8Executor17runFunctionAsMainEPN4llvm8FunctionEiPPcS5_+0x7c0)[0x5644e2410770]
klee(main+0x30ec)[0x5644e23e212c]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f1001dfeb97]
klee(_start+0x2a)[0x5644e23edbda]
Aborted (core dumped)
```

@Mqsun thanks for the report. Unfortunately, I cannot compile your example. Did you miss something?

```
clang-6.0 -emit-llvm -c -g bug_1257.c
bug_1257.c:1:2: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
(c) {
 ^
bug_1257.c:1:4: error: expected ';' after top level declarator
(c) {
   ^
   ;
```

@MartinNowack Thank you for your replying. Did you miss 'b;' in line 1 , and miss 'a' before '(c) {' in line 2.
The C program starts with
```C
b;
a(c); {
  if (!c)
    abort();
}
```
I could compile the example successfully in my computer.
```
clang -I ~/klee/klee/include/ -emit-llvm -c -g test.c 
test.c:1:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
b;
^
test.c:2:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
a(c) {
^
test.c:4:5: warning: implicitly declaring library function 'abort' with type
      'void (void) __attribute__((noreturn))' [-Wimplicit-function-declaration]
    abort();
    ^
test.c:4:5: note: include the header <stdlib.h> or explicitly provide a
      declaration for 'abort'
test.c:5:1: warning: control may reach end of non-void function [-Wreturn-type]
}
^
test.c:6:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
main() {
^
test.c:9:3: warning: implicit declaration of function 'klee_make_symbolic' is
      invalid in C99 [-Wimplicit-function-declaration]
  klee_make_symbolic(&d, sizeof(d), "__sym___VERIFIER_nondet_short");
  ^
test.c:18:8: warning: expression result unused [-Wunused-value]
    f[d];
    ~ ~^
test.c:19:16: warning: expression result unused [-Wunused-value]
    g[b - 1 - d];
    ~ ~~~~~~~~~^
8 warnings generated.
```

--- a/lib/Solver/STPBuilder.cpp
+++ b/lib/Solver/STPBuilder.cpp
@@ -436,12 +436,8 @@ ExprHandle STPBuilder::constructSDivByConstant(ExprHandle expr_n, unsigned width
   if (!hashed) {
     // STP uniques arrays by name, so we make sure the name is unique by
     // using the size of the array hash as a counter.
-    std::string unique_id = llvm::itostr(_arr_hash._array_hash.size());
-    unsigned const uid_length = unique_id.length();
-    unsigned const space = (root->name.length() > 32 - uid_length)
-                               ? (32 - uid_length)
-                               : root->name.length();
-    std::string unique_name = root->name.substr(0, space) + unique_id;
+    std::string unique_id = llvm::utostr(_arr_hash._array_hash.size());
+    std::string unique_name = root->name + unique_id;
 
     array_expr = buildArray(unique_name.c_str(), root->getDomain(),
                             root->getRange());

--- a/lib/Solver/Z3Builder.cpp
+++ b/lib/Solver/Z3Builder.cpp
@@ -392,12 +392,8 @@ Z3ASTHandle Z3Builder::getInitialArray(const Array *root) {
   if (!hashed) {
     // Unique arrays by name, so we make sure the name is unique by
     // using the size of the array hash as a counter.
-    std::string unique_id = llvm::itostr(_arr_hash._array_hash.size());
-    unsigned const uid_length = unique_id.length();
-    unsigned const space = (root->name.length() > 32 - uid_length)
-                               ? (32 - uid_length)
-                               : root->name.length();
-    std::string unique_name = root->name.substr(0, space) + unique_id;
+    std::string unique_id = llvm::utostr(_arr_hash._array_hash.size());
+    std::string unique_name = root->name + unique_id;
 
     array_expr = buildArray(unique_name.c_str(), root->getDomain(),
                             root->getRange());

--- a/test/regression/2020-04-27-stp-array-names.c
+++ b/test/regression/2020-04-27-stp-array-names.c
@@ -1,0 +1,24 @@
+// RUN: %clang %s -emit-llvm %O0opt -g -c -o %t.bc
+// RUN: rm -rf %t.klee-out
+// RUN: %klee -output-dir=%t.klee-out --search=bfs --max-instructions=1000 %t.bc
+b;
+a(c) {
+  if (!c)
+    abort();
+}
+main() {
+  int e;
+  short d;
+  klee_make_symbolic(&d, sizeof(d), "__sym___VERIFIER_nondet_short");
+  b = d;
+  a(b > 0);
+  int f[b];
+  int g[b];
+  for (;;) {
+    short d;
+    klee_make_symbolic(&d, sizeof(d), "__sym___VERIFIER_nondet_short");
+    a(d >= 0 && d < b);
+    f[d];
+    g[b - 1 - d];
+  }
+}

